# HW2_대구맛집
2022236204 홍정빈

In [154]:
import pandas as pd
import numpy as np
import seaborn as sns
import re
import warnings
warnings.filterwarnings(action = 'ignore')

In [155]:
# !pip install requests

In [156]:
import json

In [157]:
import requests
url = 'https://www.daegufood.go.kr/kor/api/tasty.html?mode=json&addr=%EC%A4%91%EA%B5%AC'
#gu = [중구','동구','남구','서구','북구','수성구','달서구''달성군']
r = requests.Session().get(url)
r.status_code

200

In [158]:
r.headers['content-type']

'application/json; Charset=utf-8'

In [159]:
r.encoding

'utf-8'

In [160]:
# 텍스트 확인
# r.content
# r.test

In [161]:
# 그냥 r.json()으로 불러오면 깨진다.
# 아무래도 \n와 같은 개행문자가 포함되어 있기 때문으로 여서 strict=False 옵션을 준다.
r.json(strict=False)

{'status': 'DONE',
 'total': '181',
 'data': [{'cnt': '1',
   'OPENDATA_ID': '1694',
   'GNG_CS': '대구광역시 중구 삼덕동2가 219-1',
   'FD_CS': '한식',
   'BZ_NM': '송림식당',
   'TLNO': '053-424-3838',
   'MBZ_HR': '11:00 ~ 20:00',
   'SEAT_CNT': '80석',
   'PKPL': '10대',
   'HP': '없음',
   'PSB_FRN': '영어 ',
   'BKN_YN': '가능',
   'INFN_FCL': '불가능',
   'BRFT_YN': '불가능',
   'DSSRT_YN': '가능',
   'MNU': '복어매운탕 9,000원\r\n복어지리 9,000원\r\n복어수육 45,000원\r\n모듬회 20,000원\r\n껍질무침 15,000원',
   'SMPL_DESC': '`송림복어` 는 50년 이상 전통의 복어 요리 전문점입니다.',
   'SBW': '지하철 2호선 경대벽원역 4번 출구에서 도보로 약 232m 거리.',
   'BUS': '버스 정류장은 `경북대학교병원앞` 정류장이 가장 가깝습니다.'},
  {'cnt': '2',
   'OPENDATA_ID': '1691',
   'GNG_CS': '대구광역시 중구 삼덕동1가 21-10',
   'FD_CS': '양식',
   'BZ_NM': '빈플러스 테이스팅 스푼(TASTING SPOON)',
   'TLNO': '053-424-5925',
   'MBZ_HR': '11:00 ~ 22:00',
   'SEAT_CNT': '22석',
   'PKPL': '유료주차장',
   'HP': '없음',
   'PSB_FRN': '가능한 외국어가 없습니다.',
   'BKN_YN': '가능',
   'INFN_FCL': '불가능',
   'BRFT_YN': '불가능',
   'DSSRT_YN': '불가능',
   'MNU': '오늘의 스

## 구이름 16진수 변환

- 중구  -> eca491eab5ac  -> %EC%A4%91%EA%B5%AC
- 동구  -> eb8f99eab5ac  -> %EB%8F%99%EA%B5%AC
- 서구  -> ec849ceab5ac  -> %EC%84%9C%EA%B5%AC
- 남구  -> eb82a8eab5ac  -> %EB%82%A8%EA%B5%AC
- 북구  -> ebb681eab5ac  -> %EB%B6%81%EA%B5%AC
- 수성구  -> ec8898ec84b1eab5ac  -> %EC%88%98%EC%84%B1%EA%B5%AC
- 달서구  -> eb8bacec849ceab5ac  -> %EB%8B%AC%EC%84%9C%EA%B5%AC
- 달성군  -> eb8bacec84b1eab5b0  -> %EB%8B%AC%EC%84%B1%EA%B5%B0

In [162]:
# 구별로 json 불러오기

url = 'https://www.daegufood.go.kr/kor/api/tasty.html?mode=json&addr='
url_jung = url + '%EC%A4%91%EA%B5%AC'
url_dong = url + '%EB%8F%99%EA%B5%AC'
url_seo = url + '%EC%84%9C%EA%B5%AC'
url_nam = url + '%EB%82%A8%EA%B5%AC'
url_buk = url + '%EB%B6%81%EA%B5%AC'
url_su = url + '%EC%88%98%EC%84%B1%EA%B5%AC'
url_dal = url + '%EB%8B%AC%EC%84%9C%EA%B5%AC'
url_daln = url + '%EB%8B%AC%EC%84%B1%EA%B5%B0'

r_jung = requests.get(url_jung)
r_dong = requests.get(url_dong)
r_seo = requests.get(url_seo)
r_nam = requests.get(url_nam)
r_buk = requests.get(url_buk)
r_su = requests.get(url_su)
r_dal = requests.get(url_dal)
r_daln = requests.get(url_daln)

print(r_jung.status_code)
print(r_dong.status_code)
print(r_seo.status_code)
print(r_nam.status_code)
print(r_buk.status_code)
print(r_su.status_code)
print(r_dal.status_code)
print(r_daln.status_code)

json_jung = r_jung.json(strict=False)
json_dong = r_dong.json(strict=False)
json_seo = r_seo.json(strict=False)
json_nam = r_nam.json(strict=False)
json_buk = r_buk.json(strict=False)
json_su = r_su.json(strict=False)
json_dal = r_dal.json(strict=False)
json_daln = r_daln.json(strict=False)

200
200
200
200
200
200
200
200


## 과제1.

### 대구의 7개 구(중구, 동구, 서구, 남구, 북구, 수성구, 달서구)과 1개 군(달성군)의 맛집은 각각 몇 개인가?

In [163]:
print('중구 맛집 수 : ', json_jung['total'])
print('동구 맛집 수 : ', json_dong['total'])
print('서구 맛집 수 : ', json_seo['total'])
print('남구 맛집 수 : ', json_nam['total'])
print('북구 맛집 수 : ', json_buk['total'])
print('수성구 맛집 수 : ', json_su['total'])
print('달서구 맛집 수 : ', json_dal['total'])
print('달성군 맛집 수 : ', json_daln['total'])

print('대구시 전체 맛집 수 : ',
      int(json_jung['total'])+
      int(json_dong['total'])+
      int(json_seo['total'])+
      int(json_nam['total'])+
      int(json_buk['total'])+
      int(json_su['total'])+
      int(json_dal['total'])+
      int(json_daln['total']))

중구 맛집 수 :  181
동구 맛집 수 :  137
서구 맛집 수 :  63
남구 맛집 수 :  57
북구 맛집 수 :  95
수성구 맛집 수 :  131
달서구 맛집 수 :  168
달성군 맛집 수 :  93
대구시 전체 맛집 수 :  925


# KoNLPy

In [164]:
# 구별 json 파일들을 하나의 리스트로 만들어서 사용하기 쉽게 한다.
dg_mat = [json_jung, json_dong, json_seo, json_nam, json_buk, json_su, json_dal, json_daln]

In [165]:
#!pip install JPype1-0.5.7-cp27-none-win_amd64.whl

In [166]:
#!pip install konlpy

In [167]:
from konlpy.tag import Kkma
from konlpy.utils import pprint # pretty print
kkma = Kkma()

In [168]:
# 테스트
pprint(kkma.sentences(u'네, 안녕하세요. 반갑습니다.'))
# 한글을 글자 깨짐 없이 사용하고 싶은 경우에는 유니코드 문자열(u'')을 사용.
# 자세한 내용 : https://ponyozzang.tistory.com/258

['네, 안녕하세요.', '반갑습니다.']


In [169]:
# 테스트
pprint(kkma.nouns(u'질문이나 건의사항은 깃헙 이슈 트래커에 남겨주세요.'))

['질문', '건의', '건의사항', '사항', '깃헙', '이슈', '트래커']


In [170]:
# 테스트
pprint(kkma.pos(u'질문이나 건의사항은 깃헙 이슈 트래커에 남겨주세요.'))

[('질문', 'NNG'),
 ('이나', 'JC'),
 ('건의', 'NNG'),
 ('사항', 'NNG'),
 ('은', 'JX'),
 ('깃헙', 'UN'),
 ('이슈', 'NNG'),
 ('트래커', 'UN'),
 ('에', 'JKM'),
 ('남기', 'VV'),
 ('어', 'ECS'),
 ('주', 'VXV'),
 ('세요', 'EFN'),
 ('.', 'SF')]


In [171]:
# SMPL_DESC 위치 확인 후 출력
json_jung['data'][0]['SMPL_DESC']

'`송림복어` 는 50년 이상 전통의 복어 요리 전문점입니다.'

In [172]:
pprint(kkma.nouns(json_jung['data'][0]['SMPL_DESC']))
# nouns를 사용하니 단어가 중복되어서 나오는 경우가 있다.

['송림', '송림복어', '복어', '50', '50년', '년', '이상', '전통', '요리', '전문점']


In [173]:
kkma.tagset
# pos를 사용해서 명사('NN','NNB','NNG','NNM','NNP','NP')와 어근('XR')만 사용하도록 하자
# 이런 식으로 pos를 이용하면 따로 숫자, 특수문자 등을 처리해줄 필요도 없다.

{'EC': '연결 어미',
 'ECD': '의존적 연결 어미',
 'ECE': '대등 연결 어미',
 'ECS': '보조적 연결 어미',
 'EF': '종결 어미',
 'EFA': '청유형 종결 어미',
 'EFI': '감탄형 종결 어미',
 'EFN': '평서형 종결 어미',
 'EFO': '명령형 종결 어미',
 'EFQ': '의문형 종결 어미',
 'EFR': '존칭형 종결 어미',
 'EP': '선어말 어미',
 'EPH': '존칭 선어말 어미',
 'EPP': '공손 선어말 어미',
 'EPT': '시제 선어말 어미',
 'ET': '전성 어미',
 'ETD': '관형형 전성 어미',
 'ETN': '명사형 전성 어미',
 'IC': '감탄사',
 'JC': '접속 조사',
 'JK': '조사',
 'JKC': '보격 조사',
 'JKG': '관형격 조사',
 'JKI': '호격 조사',
 'JKM': '부사격 조사',
 'JKO': '목적격 조사',
 'JKQ': '인용격 조사',
 'JKS': '주격 조사',
 'JX': '보조사',
 'MA': '부사',
 'MAC': '접속 부사',
 'MAG': '일반 부사',
 'MD': '관형사',
 'MDN': '수 관형사',
 'MDT': '일반 관형사',
 'NN': '명사',
 'NNB': '일반 의존 명사',
 'NNG': '보통명사',
 'NNM': '단위 의존 명사',
 'NNP': '고유명사',
 'NP': '대명사',
 'NR': '수사',
 'OH': '한자',
 'OL': '외국어',
 'ON': '숫자',
 'SE': '줄임표',
 'SF': '마침표, 물음표, 느낌표',
 'SO': '붙임표(물결,숨김,빠짐)',
 'SP': '쉼표,가운뎃점,콜론,빗금',
 'SS': '따옴표,괄호표,줄표',
 'SW': '기타기호 (논리수학기호,화폐기호)',
 'UN': '명사추정범주',
 'VA': '형용사',
 'VC': '지정사',
 'VCN': "부정 지정사, 형용사 '아니다'",
 'VC

---
---
---

# 과제2.
### 오픈 api에 포함된 "설명"(SMPL_DESC)을 분석하여, 맛집 설명에서 가장 유의미하게 많이 등장하는 5개 단어를 제시하라

## 2-1. kkma.pos 방법

In [174]:
# dg_mat[0]['data'][180]['SMPL_DESC']
# dg_mat[0] 은 중구, [180]은 181번째 가게를 뜻한다.

In [175]:
# 일단 모든 구들의 SMPL_DESC 를 하나로 합쳐주자.
dg_smpl = []
for i in range(0, 8):
    for j in range(0, int(dg_mat[i]['total'])) :
        dg_smpl.append(dg_mat[i]['data'][j]['SMPL_DESC']) # 리스트 형태로 합쳐준다.

In [176]:
# print(df_smpl)

In [177]:
# 위 방식만 하면 SMPL_DESC들이 각각의 인덱스로 들어갈 뿐 아직 전체가 하나의 문장이 되지는 않았기 때문에
# join : 리스트인 dg_mnu를 하나의 str타입으로 합쳐주고 그 사이엔 ', '를 넣는다.
dg_smpl = ', '.join(dg_smpl) 

In [178]:
# kkma.pos를 이용해서 토큰화 한 것을 변수에 담아준다.
# 변수에 담지 않고 for, if문을 돌릴 경우 kkma.pos 함수 호출을 매번 하게 되어서 어마어마한 시간이 걸린다.
dg_smpl_token = kkma.pos(dg_smpl)

In [179]:
# kkma.pos를 사용해 명사와 어근만 추출하자

dg_smpl_token_N = []
for i in range(0, len(dg_smpl_token)) :
    if dg_smpl_token[i][1] == 'NN' :
        dg_smpl_token_N.append(dg_smpl_token[i][0])
    elif dg_smpl_token[i][1] == 'NNB' :
        dg_smpl_token_N.append(dg_smpl_token[i][0])
    elif dg_smpl_token[i][1] == 'NNG' :
        dg_smpl_token_N.append(dg_smpl_token[i][0])
    elif dg_smpl_token[i][1] == 'NNM' :
        dg_smpl_token_N.append(dg_smpl_token[i][0])
    elif dg_smpl_token[i][1] == 'NNP' :
        dg_smpl_token_N.append(dg_smpl_token[i][0])
    elif dg_smpl_token[i][1] == 'NP' :
        dg_smpl_token_N.append(dg_smpl_token[i][0])
    elif dg_smpl_token[i][1] == 'XR' :
        dg_smpl_token_N.append(dg_smpl_token[i][0])

In [180]:
# print(dg_smpl_token_N)

In [181]:
# 토큰별(단어별) 카운팅!
from collections import Counter

dg_smpl_token_cnt = Counter(dg_smpl_token)
dg_smpl_token_cnt.most_common(10)

[((',', 'SP'), 1420),
 (('ㄴ', 'ETD'), 1293),
 (('.', 'SF'), 1153),
 (('하', 'XSV'), 1100),
 (('는', 'ETD'), 853),
 (('이', 'VCP'), 843),
 (('ㅂ니다', 'EFN'), 686),
 (('전문점', 'NNG'), 578),
 (('을', 'JKO'), 560),
 (('의', 'JKG'), 451)]

### 한국어는 토큰화 단계에서 조사, 접속사 등을 제거하므로 따로 정해진 불용어(stopwords)가 없다.

### 또한, 한 글자만 있는 것을 불용어로 지정해 제거할 경우 '찜', '탕', '회' 등도 제거되기 때문에 그대로 두기로 했다.

In [182]:
# 만약 불용어를 제거하고 싶다면 아래 코드를 사용한다.

'''
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

stop_words = ['은 는 이 가 등 전문점 전문 음식점 음식 식당 가게 요리 맛 맛집 집 대구 곳 년']
word_tokens = word_tokenize(dg_smpl_token_N)
stop_words=stop_words.split(' ')

result = []
for w in word_tokens: 
    if w not in stop_words: 
        result.append(w) 
        
print(word_tokens,'\n') 
print(result)
'''

"\nfrom nltk.corpus import stopwords\nfrom nltk.tokenize import word_tokenize\n\nstop_words = ['은 는 이 가 등 전문점 전문 음식점 음식 식당 가게 요리 맛 맛집 집 대구 곳 년']\nword_tokens = word_tokenize(dg_smpl_token_N)\nstop_words=stop_words.split(' ')\n\nresult = []\nfor w in word_tokens: \n    if w not in stop_words: \n        result.append(w) \n        \nprint(word_tokens,'\n') \nprint(result)\n"

In [183]:
# 토큰별(단어별) 카운팅!
from collections import Counter

dg_smpl_token_N_cnt = Counter(dg_smpl_token_N)
dg_smpl_token_N_cnt.most_common(20)

[('전문점', 578),
 ('수', 330),
 ('요리', 245),
 ('년', 211),
 ('맛', 205),
 ('은', 175),
 ('집', 159),
 ('곳', 145),
 ('등', 142),
 ('음식점', 140),
 ('다양', 140),
 ('위치', 133),
 ('음식', 113),
 ('전통', 109),
 ('사용', 107),
 ('한우', 106),
 ('대구', 105),
 ('찜', 100),
 ('식당', 97),
 ('갈비', 96)]

### 불용어를 지정해주지 않아 불필요한 토큰들이 많이 보이지만
### kkma.pos를 사용해서 도출한 유의미한 단어 5가지는 아래와 같다.

## 1. 다양
## 2. 전통
## 3. 한우
## 4. 찜
## 5. 갈비

※ '전문점'이 중요한 단어라고 생각할 수 있지만, 당연하게도 식당이 스스로를 '비전문점'이라고 홍보하는 경우는 없기 때문에 '전문점'은 빼기로 했다.

---

## 2-2. hannanum.nouns 방법

In [184]:
from konlpy.tag import Hannanum
hannanum = Hannanum()

pprint(hannanum.nouns(json_jung['data'][0]['SMPL_DESC']))
# hannanum.nouns는 kkma.nouns와는 다르게 단어가 중복되지 않는다!

['`송림복어`', '50년', '이상', '전통', '복어', '요리', '전문점']


In [185]:
dg_smpl_token_han = hannanum.nouns(dg_smpl)

In [186]:
dg_smpl_token_han_cnt = Counter(dg_smpl_token_han)
dg_smpl_token_han_cnt.most_common(30)

[('전문점', 558),
 ('수', 333),
 ('곳', 142),
 ('맛', 136),
 ('등', 129),
 ('요리', 124),
 ('음식점', 117),
 ('사용', 109),
 ('다양한', 102),
 ('위치한', 101),
 ('전통', 97),
 ('판매', 82),
 ('유명', 76),
 ('음식', 70),
 ('대구', 70),
 ('은', 65),
 ('한우', 61),
 ('운영', 60),
 ('이상', 59),
 ('식', 58),
 ('국내산', 49),
 ('집', 48),
 ('신선한', 46),
 ('내', 46),
 ('종류', 43),
 ('재료', 42),
 ('다양', 38),
 ('칼국수', 37),
 ('건강', 35),
 ('돼지고기', 35)]

### hannanum.nouns를 사용해서 도출한 유의미한 단어 5가지는 아래와 같다.
## 1. 다양한
## 2. 전통
## 3. 한우
## 4. 국내산
## 5. 신선한

kkma.pos와 hannanum.nouns 두 방법 중 hannanum.nouns의 결과를 채택하기로 했다.
그 이유는 kkma.pos의 '찜', '갈비'는 정확히 어떤 음식을 가리키는지(갈비찜, 아구찜, 돼지갈비, 소갈비 등) 알 수 없기 때문이다.

위 단어들을 유의미하다고 판단한 이유는 아래와 같다.

1. **다양한** : 메뉴의 다양성을 홍보하기 위해 사용된 것으로 보인다. 현대사회에 들어서면서 다양성 존중에 대한 인식은 사회 전반에 자리잡았다. 한국 역시 과거 집단주의 사회에서 점차 개인주의 사회로 변화하고 있으며 이에 따라 '개인화'와 '개인 맞춤형'은 시대의 트렌드로 떠올랐다. 이는 사회가 개인의 취사선택을 보다 중요하게 생각하고 있고 더 많은 이들의 needs를 충족시키려 한다는 것을 시사한다. 따라서 식당 설명란에 있는 '다양한'이라는 키워드는 식당을 홍보하기에 충분히 유의미한 단어라고 볼 수 있다.


2. **전통** : 식당의 전통과 역사를 강조하는 단어다. 전통있는 식당은 그만큼 오래 가게를 운영하며 노하우가 축적되고 단골고객도 많을 것이기 때문에 중요한 단어라고 생각된다. 다만, '전통식당'은 요즘 트렌드를 따라가지 못한다는 느낌을 받을 수도 있다. 그러나 실제 MZ 세대들은 이른바 '할매입맛'을 존중하고 더 나아가 그러한 세대를 '할매니얼' 세대라고 부를 만큼 옛스러운 것을 또다른 개성과 취향이라고 생각하는 분위기이기 때문에 전통성을 강조한 식당도 MZ세대의 주목을 충분히 받고 있는 것으로 보인다.


3. **한우** : 1990년대에 시작된 신토불이 운동과 더불어 2003년과 2008년의 미국 광우병 파동으로 인해 한국인의 한우에 대한 관심은 과거와 비교할 수 없이 커졌다. 또한 본격적으로 한우를 브랜딩화 하고 국가와 지자체가 나서서 체계적으로 관리하면서 한우는 '명품'의 이미지를 가질 수 있게 되었다. 때문에 맛집 리스트 중 한우를 판매하는 식당이 많은 것은 어쩌면 당연한 일일 것이다.


4. **국내산** : 한우와 마찬가지로 신토불이 운동을 통해 국내산에 대한 한국인들의 애정은 더욱 커졌다. 특히 현지에서 나는 식자재일 수록 유통과정이 짧기 때문에 더 신선할 수 있다는 생각에 많은 이들이 국내산을 찾고있다. 또한 원산지표시법이 강화되고 정부의 관리 아래에 국내산 농산물에 대한 품질인증과 철저한 이력추적 덕분에 국내산 식품은 믿고 먹을 수 있다는 인식이 강해졌다. 따라서 대구시 맛집들 중에서도 국내산 재료를 사용하는 맛집이 많은 것으로 보인다.


5. **신선한** : 식재료의 신선도가 맛집 선정에 있어서 중요한 이유는 최근 식품 유통기업들의 서비스를 살펴보면 알 수 있다. 먼저, 마켓컬리는 풀콜드체인 시스템을 통해 식재료를 신선한 상태로 하룻밤만에 배송하는 서비스를 제공하고 있으며, 이를 통해 1조 유니콘 기업으로 성장할 수 있었다. 이에 질세라 쿠팡에서도 로켓프레시 서비스를 새롭게 도입해 신선한 식재료를 고객들에게 배송하고 있다. 이러한 사례들을 통해 식재료의 신선도는 맛집 선정에 있어서 충분히 유의미한 단어라고 볼 수 있다.

***
*****
- - -

# 과제3.
### 오픈 api에 포함된 "메뉴"(MNU)를 분석하여, 대구 맛집들이 제공하는 가장 흔한 메뉴 3개를 제시하라. 그리고 이들의 평균 가격 또한 함께 제시하라.

## 3-1. kkma.pos 방법

In [187]:
dg_mat[0]['data'][0]['MNU']

'복어매운탕 9,000원\r\n복어지리 9,000원\r\n복어수육 45,000원\r\n모듬회 20,000원\r\n껍질무침 15,000원'

In [188]:
# 메뉴만 추출
dg_mnu = []
for i in range(0, 8):
    for j in range(0, int(dg_mat[i]['total'])) :
        dg_mnu.append(dg_mat[i]['data'][j]['MNU'])

In [189]:
len(dg_mnu)
# 대구시 전체 맛집 수와 일치한다.

925

In [190]:
# ''.join() : dg_mnu 리스트의 인덱스들을 하나의 str타입으로 합쳐주고 그 사이엔 ' ' 안의 값을 넣는다.
dg_mnu_str = ' '.join(dg_mnu) 

In [191]:
dg_mnu_str = dg_mnu_str.replace('\r\n',' ') # \r\n는 공백으로 대체해 없앤다.

In [192]:
# 나중에 가격 분석할 때 방해되는 문자들은 제거한다.
dg_mnu_str = dg_mnu_str.replace(',','') 
dg_mnu_str = dg_mnu_str.replace('.','')
dg_mnu_str = dg_mnu_str.replace('1인분','')
dg_mnu_str = dg_mnu_str.replace('100g','')
dg_mnu_str = dg_mnu_str.replace('1+','')

In [193]:
print(dg_mnu_str[:200]) # 200글자만 확인

복어매운탕 9000원 복어지리 9000원 복어수육 45000원 모듬회 20000원 껍질무침 15000원 오늘의 스테이크 19000원 새우오일 파스타 13000원 명란크림 파스타 14000원 로제새우 파스타 14000원 쌀떡볶이+납작만두 (섞어 )  4000원 쌀떡볶이 4000원 납작만두 4000원 순대 4000원 런치 토시살스테이크 코스 35000원 디너 


In [194]:
dg_mnu_token = kkma.pos(dg_mnu_str)

dg_mnu_token_N = []
for i in range(0, len(dg_mnu_token)) :
    if dg_mnu_token[i][1] == 'NNG' :
        dg_mnu_token_N.append(dg_mnu_token[i][0])

In [195]:
# 토큰별(단어별) 카운팅!
dg_mnu_token_N_cnt = Counter(dg_mnu_token_N)
dg_mnu_token_N_cnt.most_common(20)

[('원', 948),
 ('찜', 305),
 ('한우', 303),
 ('갈비', 303),
 ('정식', 253),
 ('불고기', 250),
 ('인', 238),
 ('밥', 215),
 ('구이', 200),
 ('회', 199),
 ('소', 181),
 ('돼지', 163),
 ('복', 162),
 ('오리', 145),
 ('탕', 137),
 ('양념', 137),
 ('수육', 134),
 ('모듬', 131),
 ('전골', 127),
 ('대', 123)]

### kkma.pos를 사용해 분석한 결과, 가장 흔한 메뉴 3가지는 아래와 같다.
## 1. 찜 : 305회
## 2. 한우 : 303회
## 3. 갈비 : 303회

---

## 3-2. hannanum.nouns 방법

In [196]:
len(dg_mnu_str)

107198

In [197]:
# 문자열이 너무 길어서 그런지 hannanum.nouns가 돌아가지 않아서 분리해서 실행시킨다.
dg_mnu_han1 = hannanum.nouns(dg_mnu_str[:30000])
dg_mnu_han2 = hannanum.nouns(dg_mnu_str[30000:60000])
dg_mnu_han3 = hannanum.nouns(dg_mnu_str[60000:90000])
dg_mnu_han4 = hannanum.nouns(dg_mnu_str[90000:])

In [198]:
dg_mnu_han = dg_mnu_han1 + dg_mnu_han2 + dg_mnu_han3 + dg_mnu_han4
dg_mnu_han_str = ' '.join(dg_mnu_han)

In [199]:
p = re.compile('[가-웍|웓-힣]+') # 정규표현식을 이용해 '원'을 제외한 한글만 받음. 
dg_mnu_han = p.findall(dg_mnu_han_str)

In [200]:
result_han_cnt = Counter(dg_mnu_han)
result_han_cnt.most_common(20)

[('대', 120),
 ('인', 100),
 ('소', 88),
 ('중', 86),
 ('한우', 68),
 ('인이상', 60),
 ('점심특선', 55),
 ('수육', 47),
 ('불고기', 47),
 ('삼계탕', 46),
 ('정식', 45),
 ('육회', 45),
 ('갈비살', 43),
 ('파스타', 41),
 ('생삼겹살', 38),
 ('코스', 36),
 ('물회', 30),
 ('샐러드', 29),
 ('스테이크', 28),
 ('삼겹살', 28)]

### hannanum.nouns을 사용해 분석한 결과, 가장 흔한 메뉴 3가지는 아래와 같다.
## 1. 한우 : 68회
## 2. 수육 : 47회
## 3. 불고기 : 47회

---

## 3-3. 정규표현식 방법
메뉴의 경우 조사, 접속사 등이 없으므로 그냥 정규식을 이용해 메뉴이름만 추출할 수도 있다.

In [201]:
p = re.compile('[가-웍|웓-힣]+') # 정규표현식을 이용해 '원'을 제외한 한글만 받음. 
result_normal = p.findall(dg_mnu_str) # findall()함수를 이용해 리스트로 추출.

In [202]:
# 토큰별(단어별) 카운팅!
dg_mnu__normal_cnt = Counter(result_normal)
dg_mnu__normal_cnt.most_common(20)

[('인', 155),
 ('대', 120),
 ('중', 86),
 ('소', 82),
 ('한우', 68),
 ('인이상', 67),
 ('점심특선', 56),
 ('수육', 50),
 ('불고기', 50),
 ('정식', 49),
 ('삼계탕', 46),
 ('육회', 45),
 ('파스타', 43),
 ('갈비살', 43),
 ('생삼겹살', 39),
 ('코스', 36),
 ('인분', 33),
 ('물회', 30),
 ('아구찜', 30),
 ('삼겹살', 29)]

### 정규표현식을 사용해 분석한 결과, 가장 흔한 메뉴 3가지는 아래와 같다.
## 1. 한우 : 68회
## 2. 수육 : 50회
## 3. 불고기 : 50회

kkma.pos를 사용할 경우 '한우'는 303번 나오지만, hannanum.nouns와 정규표현식에서는 68번만 나온다.<br>
이렇게 차이가 큰 이유는  hannanum.nouns와 정규표현식을 사용할 경우 붙어있는 한글들을 한 덩어리로 보고 '한우등심', '한우갈비' 등은 '한우'로 인식하지 못하기 때문이다.


어떤 결과를 채택하든 해당 단어가 정확히 어떤 음식을 가리키는지는 알 수 없다.<br>
**따라서 hannanum.nouns와 정규표현식의 결과가 유사하기 때문에 이를 따르기로 했다.**


### 3-3-1. 이름에 '한우'가 들어간 메뉴의 평균 가격

In [283]:
dg_mnu_str = dg_mnu_str.replace(' ', '')

In [284]:
i = 0
x_index = []
w_index = []
while i < len(dg_mnu_str):
    x = dg_mnu_str.find('한우', i)
    x_index.append(dg_mnu_str.find('한우', i))
    i = x + 1
    w = dg_mnu_str.find('원', i)
    w_index.append(dg_mnu_str.find('원', i))
    
    
    if x == -1 :
        break   

In [285]:
print(len(x_index))
print(len(w_index))

320
320


In [286]:
hanwoo = []
for i in range(0,len(x_index)) :
    hanwoo.append(dg_mnu_str[x_index[i]:w_index[i]+1])

In [287]:
# hanwoo
# '한우'가 들어간 요리 종류가 너무 다양하고, g에 따라 가격이 달라진다.
# 심지어 같은 메뉴라도 가게의 네이밍 방법에 따라 이름이 또 달라진다.
# 이걸 어떻게 평균을 구하라는....

In [288]:
# 일단 ( ~g)을 다 없애주자. 먼저 다시 str로 만든다.
hanwoo_str = ' '.join(hanwoo)

In [290]:
i = 0
hs_index=[]
he_index=[]
while i < len(hanwoo_str):
    hs = hanwoo_str.find('(', i)
    hs_index.append(hanwoo_str.find('(', i))
    i = hs + 1
    he = hanwoo_str.find(')', i)
    he_index.append(hanwoo_str.find(')', i))
    
    if hs == -1 :
        break   

In [291]:
print(len(hs_index),len(he_index))

179 179


In [292]:
hanwoo_str2 = hanwoo_str
for i in range(len(hs_index)) :
    hanwoo_str2 = hanwoo_str2.replace(hanwoo_str2[hs_index[i]:he_index[i]+1],' '*len(hanwoo_str2[a_index[i]:b_index[i]+1]))

In [293]:
hanwoo_str3 = hanwoo_str2.replace(' ', '')
hanwoo_list = hanwoo_str3.split('원')

In [294]:
hanwoo_list

['한우채끝등심코스75000',
 '한우불고기정식11000',
 '한우갈비탕13000',
 '한우갈비찜정식19000',
 '한우한마리45000',
 '한우차돌박이12000',
 '한우등심22000',
 '한우꽃갈비25000',
 '한우갈빗살19000',
 '한우특수부위25000',
 '한우육회25000',
 '한우찜갈비30000',
 '한우후쿠오카모츠나베47000~48000',
 '한우우설야카니쿠56000',
 '한우숯불스테이크야키니쿠45900',
 '한우부채살야키니쿠44900',
 '한우특갈비20000',
 '한우육회19000',
 '한우생모둠코스38000',
 '한우갈비살코스34000',
 '한우탕11000',
 '한우)14000',
 '한우)17000',
 '한우)17000',
 '한우)17000',
 '한우)22000',
 '한우)22000',
 '한우)11000',
 '한우)30000',
 '한우)45000',
 '한우)55000',
 '한우)55000',
 '한우)60000',
 '한우꽃등심15000',
 '한우육회20000',
 '한우불고기10000',
 '한우소전골7000',
 '한우찜갈비30000',
 '한우찜갈비30000',
 '한우찜갈비30000',
 '한우찜갈비30000',
 '한우찜갈비30000',
 '한우찜갈비30000',
 '한우찜갈비30000',
 '한우찜갈비30000',
 '한우갈비살22000',
 '한우등심20000',
 '한우버섯불고기14000',
 '한우육회28000',
 '한우안창살26000',
 '한우쟁반70000',
 '한우갈비탕15000수육30000',
 '한우뭉티기25000',
 '한우안창살33000',
 '한우생갈비29000',
 '한우양념갈비29000',
 '한우불고기16000',
 '한우육개장9000',
 '한우갈비뼈탕9000',
 '한우갈비살14000',
 '한우꽃살17000',
 '한우등심12000',
 '한우안창살22000',
 '한우육회20000',
 '한우모듬95000',
 '한우등심모듬80000',


In [295]:
# 마지막에 붙어있는 빈 리스트도 제거 
hanwoo_list.pop()

''

In [296]:
# 붙어있는 메뉴 분리해서 다시 넣기(수육은 버림)
hanwoo_list.index('한우갈비탕15000수육30000')

51

In [297]:
hanwoo_list[51] = '한우갈비탕15000'

In [298]:
# hanwoo_list 

In [324]:
h_price=[]
for i in range(len(hanwoo_list)) :
    p_num = re.compile('[0-9]+')
    h_price.append(p_num.findall(hanwoo_list[i]))
#h_price

In [325]:
#['47000', '48000'] 같이 한 리스트에 2개 가격이 들어간거 처리
for i in range(len(h_price)) :
    if len(h_price[i]) == 2 :
        h_price.insert(i+1, h_price[i][1])
        h_price[i] = h_price[i][0]
    elif len(h_price[i]) == 3 :
        h_price.insert(i+1, h_price[i][2])
        h_price.insert(i+1, h_price[i][1])
        h_price[i] = h_price[i][0]

In [326]:
#h_price

In [327]:
# 리스트를 문자열로 바꾸고 대괄호 및 따옴표 제거
for i in range(len(h_price)) :
    h_price[i] = str(h_price[i]).replace('[', '')
    h_price[i] = h_price[i].replace(']', '')
    h_price[i] = h_price[i].replace("'", '')

In [328]:
for i in range(len(h_price)) : # 100단위 이하 숫자는 제거
    if len(h_price[i]) <= 3 :
        del h_price[i]

In [333]:
for i in range(len(h_price)) :
    h_price[i] = int(h_price[i]) # 평균 계산을 위해 숫자형으로

In [334]:
# 이름에 '한우'가 들어간 메뉴의 평균 가격
print(sum(h_price)/len(h_price))

23060.43613707165


## '한우'의 평균 가격은 23060원이다.
### 그럼 이제 '한우'가 들어간 메뉴 중에서도 가장 많은 3가지 메뉴에 대한 평균 가격을 각각 구햅자

In [344]:
hanwoo = ",".join(hanwoo) 
p = re.compile('[가-웍|웓-힣]+') # 정규표현식을 이용해 '원'을 제외한 한글만 받음. 
result_h = p.findall(hanwoo) 

In [345]:
hanwoo_top = Counter(result_h)
hanwoo_top.most_common(10)

[('한우갈비살', 29),
 ('한우육회', 22),
 ('한우', 16),
 ('한우안창살', 13),
 ('한우등심', 12),
 ('한우불고기', 10),
 ('한우찜갈비', 9),
 ('한우국밥', 8),
 ('한우소고기국밥', 8),
 ('한우등심구이', 7)]

### 한우가 들어간 메뉴 중에서도 가장 많은 3가지는 아래와 같다.
## 1. 한우갈비살
## 2. 한우육회
## 3. 한우안창살

In [354]:
hanwoo_galbi=[]
for i in range(len(hanwoo_list)) :
    if '한우갈비살' in hanwoo_list[i] :
        hanwoo_galbi.append(hanwoo_list[i])

In [355]:
hanwoo_galbi

['한우갈비살코스34000',
 '한우갈비살22000',
 '한우갈비살14000',
 '한우갈비살25000',
 '한우갈비살19000',
 '한우갈비살16000',
 '한우갈비살19000',
 '한우갈비살14800',
 '한우갈비살29000',
 '한우갈비살25000',
 '한우갈비살22000',
 '한우갈비살17500',
 '한우갈비살18000',
 '한우갈비살/양념19000',
 '한우갈비살/한우등심70000',
 '한우갈비살/한우등심102000',
 '한우갈비살20000',
 '한우갈비살18000',
 '한우갈비살특수부위22500',
 '한우갈비살26000',
 '한우갈비살21000',
 '한우갈비살21000',
 '한우갈비살20000',
 '한우갈비살20000',
 '한우갈비살28000',
 '한우갈비살구이22000',
 '한우갈비살25000',
 '한우갈비살19000',
 '한우갈비살18000',
 '한우갈비살18000',
 '한우갈비살18000',
 '한우갈비살정식30000',
 '한우갈비살17000']

In [357]:
h_galbi_price=[]
for i in range(len(hanwoo_galbi)) :
    p_num = re.compile('[0-9]+')
    h_galbi_price.append(p_num.findall(hanwoo_galbi[i]))
h_galbi_price

[['34000'],
 ['22000'],
 ['14000'],
 ['25000'],
 ['19000'],
 ['16000'],
 ['19000'],
 ['14800'],
 ['29000'],
 ['25000'],
 ['22000'],
 ['17500'],
 ['18000'],
 ['19000'],
 ['70000'],
 ['102000'],
 ['20000'],
 ['18000'],
 ['22500'],
 ['26000'],
 ['21000'],
 ['21000'],
 ['20000'],
 ['20000'],
 ['28000'],
 ['22000'],
 ['25000'],
 ['19000'],
 ['18000'],
 ['18000'],
 ['18000'],
 ['30000'],
 ['17000']]

In [360]:
for i in range(len(h_galbi_price)) :
    h_galbi_price[i] = str(h_galbi_price[i]).replace('[', '')
    h_galbi_price[i] = h_galbi_price[i].replace(']', '')
    h_galbi_price[i] = h_galbi_price[i].replace("'", '')
    h_galbi_price[i] = int(h_galbi_price[i])

In [362]:
print(sum(h_galbi_price)/len(h_galbi_price))

25145.454545454544


In [363]:
hanwoo_youk=[]
for i in range(len(hanwoo_list)) :
    if '한우육회' in hanwoo_list[i] :
        hanwoo_youk.append(hanwoo_list[i])
        
h_youk_price=[]
for i in range(len(hanwoo_youk)) :
    p_num = re.compile('[0-9]+')
    h_youk_price.append(p_num.findall(hanwoo_youk[i]))
    
for i in range(len(h_youk_price)) :
    h_youk_price[i] = str(h_youk_price[i]).replace('[', '')
    h_youk_price[i] = h_youk_price[i].replace(']', '')
    h_youk_price[i] = h_youk_price[i].replace("'", '')
    h_youk_price[i] = int(h_youk_price[i])

print(sum(h_youk_price)/len(h_youk_price))

20383.870967741936


In [ ]:
hanwoo_youk=[]
for i in range(len(hanwoo_list)) :
    if '한우안창살' in hanwoo_list[i] :
        hanwoo_youk.append(hanwoo_list[i])
        
h_youk_price=[]
for i in range(len(hanwoo_youk)) :
    p_num = re.compile('[0-9]+')
    h_youk_price.append(p_num.findall(hanwoo_youk[i]))
    
for i in range(len(h_youk_price)) :
    h_youk_price[i] = str(h_youk_price[i]).replace('[', '')
    h_youk_price[i] = h_youk_price[i].replace(']', '')
    h_youk_price[i] = h_youk_price[i].replace("'", '')
    h_youk_price[i] = int(h_youk_price[i])

print(sum(h_youk_price)/len(h_youk_price))

### 3-3-2. 이름에 '수육'이 들어간 메뉴의 평균 가격

In [135]:
i = 0
x_index = []
w_index = []
while i < len(dg_mnu_str):
    x = dg_mnu_str.find('수육', i)
    x_index.append(dg_mnu_str.find('수육', i))
    i = x + 1
    w = dg_mnu_str.find('원', i)
    w_index.append(dg_mnu_str.find('원', i))
    
    
    if x == -1 :
        break 

In [136]:
sooyouk = []
for i in range(0,len(x_index)) :
    sooyouk.append(dg_mnu_str[x_index[i]:w_index[i]+1])

In [137]:
sooyouk = ' '.join(sooyouk)

In [138]:
#sooyouk

'수육45000원 수육10000원 수육30000원 수육/갈비볶음(中)32000원 수육10000생막창(200g)8000원 수육40000원 수육68000원 수육12000원 수육(대)20000원 수육18000원 수육12000원 수육(한우)30000원 수육(한우)45000원 수육(한우)55000원 수육(한우)55000원 수육(한우)60000원 수육15000원 수육정식8500원 수육30000원 수육15000원 수육25000원 수육15000원 수육24000원 수육(대)35000원 수육+순대+국밥2)25000원 수육22000원 수육(예약)40000원 수육35000원 수육20000원 수육백반8000원 수육20000원 수육18000원 수육30000원 수육18000원 수육17000원 수육32000원 수육22000원 수육30000원 수육20000원 수육15000원 수육전골35000원 수육10000원 수육)25000원 수육)35000원 수육15000원 수육9900원 수육45000원 수육보쌈한접시12000원 수육보쌈정식9000원 수육보쌈Set9000원 수육백반8000원 수육암뽕20000원 수육12000원 수육16000원 수육20000원 수육한접시:20000원 수육/암뽕/순대30000원 수육22000원 수육+암뽕22000원 수육20000원 수육10000원 수육:30000원 수육:40000원 수육30000원 수육50000원 수육60000원 수육+만두+튀김+밀복불고기+밀복탕/지리)35000원 수육(암뽕)12000원 수육포장22000원 수육+칼국수+밥)12000원 수육30000원 수육25000원 수육15000원 수육(중)10000(대)15000납작만두3000원 수육20000원 수육40000원 수육50000원 수육25000원 수육/복튀김/찜종류30000원 수육백반11000원 수육내장수육10000원 수육10000원 수육/양수육/도가니수육35000원 수육/도가니수육35000원 수육35000원 수육20000원 수육(중)20000원 수육백반10000원 수육20000원 수육17000원 수육20000원

In [139]:
sooyouk_list = sooyouk.split('원')

In [140]:
sooyouk_list

['수육45000',
 ' 수육10000',
 ' 수육30000',
 ' 수육/갈비볶음(中)32000',
 ' 수육10000생막창(200g)8000',
 ' 수육40000',
 ' 수육68000',
 ' 수육12000',
 ' 수육(대)20000',
 ' 수육18000',
 ' 수육12000',
 ' 수육(한우)30000',
 ' 수육(한우)45000',
 ' 수육(한우)55000',
 ' 수육(한우)55000',
 ' 수육(한우)60000',
 ' 수육15000',
 ' 수육정식8500',
 ' 수육30000',
 ' 수육15000',
 ' 수육25000',
 ' 수육15000',
 ' 수육24000',
 ' 수육(대)35000',
 ' 수육+순대+국밥2)25000',
 ' 수육22000',
 ' 수육(예약)40000',
 ' 수육35000',
 ' 수육20000',
 ' 수육백반8000',
 ' 수육20000',
 ' 수육18000',
 ' 수육30000',
 ' 수육18000',
 ' 수육17000',
 ' 수육32000',
 ' 수육22000',
 ' 수육30000',
 ' 수육20000',
 ' 수육15000',
 ' 수육전골35000',
 ' 수육10000',
 ' 수육)25000',
 ' 수육)35000',
 ' 수육15000',
 ' 수육9900',
 ' 수육45000',
 ' 수육보쌈한접시12000',
 ' 수육보쌈정식9000',
 ' 수육보쌈Set9000',
 ' 수육백반8000',
 ' 수육암뽕20000',
 ' 수육12000',
 ' 수육16000',
 ' 수육20000',
 ' 수육한접시:20000',
 ' 수육/암뽕/순대30000',
 ' 수육22000',
 ' 수육+암뽕22000',
 ' 수육20000',
 ' 수육10000',
 ' 수육:30000',
 ' 수육:40000',
 ' 수육30000',
 ' 수육50000',
 ' 수육60000',
 ' 수육+만두+튀김+밀복불고기+밀복탕/지리)35000',
 ' 수육(암뽕)12000',

In [141]:
sooyouk_list[4] = '수육10000'

In [142]:
sooyouk_list.pop()

' '

In [143]:
s_price=[]
for i in range(len(sooyouk_list)) :
    p = re.compile('[0-9]+')
    s_price.append(p.findall(sooyouk_list[i]))
#s_price

[['45000'],
 ['10000'],
 ['30000'],
 ['32000'],
 ['10000'],
 ['40000'],
 ['68000'],
 ['12000'],
 ['20000'],
 ['18000'],
 ['12000'],
 ['30000'],
 ['45000'],
 ['55000'],
 ['55000'],
 ['60000'],
 ['15000'],
 ['8500'],
 ['30000'],
 ['15000'],
 ['25000'],
 ['15000'],
 ['24000'],
 ['35000'],
 ['2', '25000'],
 ['22000'],
 ['40000'],
 ['35000'],
 ['20000'],
 ['8000'],
 ['20000'],
 ['18000'],
 ['30000'],
 ['18000'],
 ['17000'],
 ['32000'],
 ['22000'],
 ['30000'],
 ['20000'],
 ['15000'],
 ['35000'],
 ['10000'],
 ['25000'],
 ['35000'],
 ['15000'],
 ['9900'],
 ['45000'],
 ['12000'],
 ['9000'],
 ['9000'],
 ['8000'],
 ['20000'],
 ['12000'],
 ['16000'],
 ['20000'],
 ['20000'],
 ['30000'],
 ['22000'],
 ['22000'],
 ['20000'],
 ['10000'],
 ['30000'],
 ['40000'],
 ['30000'],
 ['50000'],
 ['60000'],
 ['35000'],
 ['12000'],
 ['22000'],
 ['12000'],
 ['30000'],
 ['25000'],
 ['15000'],
 ['10000', '15000', '3000'],
 ['20000'],
 ['40000'],
 ['50000'],
 ['25000'],
 ['30000'],
 ['11000'],
 ['10000'],
 ['10000'],


In [144]:
# 전처리
s_price.index(['2', '25000'])

24

In [ ]:
s_price[24] = ['25000']

In [145]:
s_price.index(['1', '160000'])

124

In [147]:
s_price[124] = ['16000']

In [148]:
#s_price

[['45000'],
 ['10000'],
 ['30000'],
 ['32000'],
 ['10000'],
 ['40000'],
 ['68000'],
 ['12000'],
 ['20000'],
 ['18000'],
 ['12000'],
 ['30000'],
 ['45000'],
 ['55000'],
 ['55000'],
 ['60000'],
 ['15000'],
 ['8500'],
 ['30000'],
 ['15000'],
 ['25000'],
 ['15000'],
 ['24000'],
 ['35000'],
 ['25000'],
 ['22000'],
 ['40000'],
 ['35000'],
 ['20000'],
 ['8000'],
 ['20000'],
 ['18000'],
 ['30000'],
 ['18000'],
 ['17000'],
 ['32000'],
 ['22000'],
 ['30000'],
 ['20000'],
 ['15000'],
 ['35000'],
 ['10000'],
 ['25000'],
 ['35000'],
 ['15000'],
 ['9900'],
 ['45000'],
 ['12000'],
 ['9000'],
 ['9000'],
 ['8000'],
 ['20000'],
 ['12000'],
 ['16000'],
 ['20000'],
 ['20000'],
 ['30000'],
 ['22000'],
 ['22000'],
 ['20000'],
 ['10000'],
 ['30000'],
 ['40000'],
 ['30000'],
 ['50000'],
 ['60000'],
 ['35000'],
 ['12000'],
 ['22000'],
 ['12000'],
 ['30000'],
 ['25000'],
 ['15000'],
 ['10000', '15000', '3000'],
 ['20000'],
 ['40000'],
 ['50000'],
 ['25000'],
 ['30000'],
 ['11000'],
 ['10000'],
 ['10000'],
 ['35

In [149]:
# 한 리스트에 2개 이상 가격이 들어간거 처리
for i in range(len(s_price)) :
    if len(s_price[i]) == 2 :
        s_price.insert(i+1, s_price[i][1])
        s_price[i] = s_price[i][0]
    elif len(s_price[i]) == 3 :
        s_price.insert(i+1, s_price[i][2])
        s_price.insert(i+1, s_price[i][1])
        s_price[i] = s_price[i][0]

In [150]:
# price

In [151]:
for i in range(len(s_price)) :
    s_price[i] = str(s_price[i]).replace('[', '')
    s_price[i] = s_price[i].replace(']', '')
    s_price[i] = s_price[i].replace("'", '')
    s_price[i] = int(s_price[i])

In [152]:
# 이름에 '수육'이 들어간 메뉴의 평균 가격
print(sum(s_price)/len(s_price))

27438.37837837838
